File: core/src/main/java/y2k/joyreactor/presenters/PostPresenter.java
package y2k.joyreactor.presenters;

import y2k.joyreactor.*;
import y2k.joyreactor.platform.Navigation;
import y2k.joyreactor.platform.Platform;
import y2k.joyreactor.services.PostService;

import java.io.File;
import java.util.List;

/**
 * Created by y2k on 28/09/15.
 */
public class PostPresenter {

    private View view;
    private PostService service;

    public PostPresenter(View view, PostService service) {
        this.view = view;
        this.service = service;
        initialize();
    }

    private void initialize() {
        view.setIsBusy(true);
        service.synchronizePostAsync(getArgumentPostId())
                .subscribe(post -> {
                    view.setIsBusy(false);
                    view.updatePostInformation(post);

                    service.getPostImages(post.id)
                            .subscribe(view::updatePostImages, Throwable::printStackTrace);
                    service.getCommentsAsync(post.id, 0)
                            .subscribe(view::updateComments, Throwable::printStackTrace);
                    service.getSimilarPosts(post.id)
                            .subscribe(view::updateSimilarPosts, Throwable::printStackTrace);
                    service.mainImagePartial(post.serverId)
                            .subscribe(partial -> {
                                if (partial.result == null) {
                                    view.updateImageDownloadProgress(partial.progress, partial.max);
                                } else {
                                    view.updatePostImage(partial.result);
                                }
                            }, Throwable::printStackTrace);
                }, Throwable::printStackTrace);
    }

    public void selectComment(int commentId) {
        service.getFromCache(getArgumentPostId())
                .flatMap(post -> service.getCommentsAsync(post.id, commentId))
                .subscribe(view::updateComments, Throwable::printStackTrace);
    }

    public void openPostInBrowser() {
        Navigation.getInstance().openBrowser("http://joyreactor.cc/post/" + getArgumentPostId());
    }

    public void saveImageToGallery() {
        view.setIsBusy(true);
        service.getFromCache(getArgumentPostId())
                .flatMap(post -> service.mainImage(post.serverId))
                .flatMap(imageFile -> Platform.Instance.saveToGallery(imageFile))
                .subscribe(_void -> {
                    view.showImageSuccessSavedToGallery();
                    view.setIsBusy(false);
                }, Throwable::printStackTrace);
    }

    private String getArgumentPostId() {
        return Navigation.getInstance().getArgumentPostId();
    }

    public void replyToComment(Comment comment) {
        // TODO:
    }

    public void replyToPost() {
        // TODO:
        Navigation.getInstance().openCreateComment();
    }

    public interface View {

        void updateComments(CommentGroup comments);

        @Deprecated
        void updatePostInformation(Post post);

        void setIsBusy(boolean isBusy);

        void showImageSuccessSavedToGallery();

        void updatePostImages(List<Image> images);

        void updateSimilarPosts(List<SimilarPost> similarPosts);

        void updatePostImage(File image);

        void updateImageDownloadProgress(int progress, int maxProgress);
    }
}

File: core/src/main/kotlin/y2k/joyreactor/presenters/PostPresenter.kt
package y2k.joyreactor.presenters

import y2k.joyreactor.*
import y2k.joyreactor.common.subscribeOnMain
import y2k.joyreactor.model.*
import y2k.joyreactor.platform.NavigationService
import y2k.joyreactor.platform.Platform
import y2k.joyreactor.services.PostService
import y2k.joyreactor.services.ProfileService
import java.io.File

/**
 * Created by y2k on 28/09/15.
 */
class PostPresenter(
    private val view: PostPresenter.View,
    private val service: PostService,
    private val userService: ProfileService,
    private val navigation: NavigationService) {

    init {
        view.setIsBusy(true)
        service
            .synchronizePostAsync(argumentPostId)
            .subscribeOnMain { post ->
                view.updatePostInformation(post)

                service
                    .getPostImages()
                    .subscribeOnMain { view.updatePostImages(it) }

                service
                    .getCommentsAsync(post.id, 0)
                    .subscribeOnMain {
                        view.updateComments(it)
                        view.setIsBusy(false)
                    }

                service
                    .getSimilarPosts(post.id)
                    .subscribeOnMain { view.updateSimilarPosts(it) }

                service
                    .mainImagePartial(post.serverId)
                    .subscribeOnMain { partial ->
                        if (partial.result == null) {
                            view.updateImageDownloadProgress(partial.progress, partial.max)
                        } else {
                            view.updatePostImage(partial.result)
                        }
                    }

                userService
                    .isAuthorized()
                    .subscribeOnMain { if (it) view.setEnableCreateComments() }
            }
    }

    fun selectComment(commentId: Long) {
        service
            .getFromCache(argumentPostId)
            .flatMap { post -> service.getCommentsAsync(post.id, commentId) }
            .subscribeOnMain { view.updateComments(it) }
    }

    fun openPostInBrowser() {
        NavigationService.instance.openBrowser("http://joyreactor.cc/post/" + argumentPostId)
    }

    fun saveImageToGallery() {
        view.setIsBusy(true)
        service.getFromCache(argumentPostId)
            .flatMap { post -> service.mainImage(post.serverId!!) }
            .flatMap { imageFile -> Platform.instance.saveToGallery(imageFile) }
            .subscribeOnMain {
                view.showImageSuccessSavedToGallery()
                view.setIsBusy(false)
            }
    }

    private val argumentPostId: String
        get() = NavigationService.instance.argumentPostId

    fun replyToComment(comment: Comment) {
        // TODO:
    }

    fun replyToPost() {
        // TODO:
        NavigationService.instance.openCreateComment()
    }

    fun showMoreImages() {
        navigation.openPostGallery()
    }

    interface View {

        fun updateComments(comments: CommentGroup)

        @Deprecated("")
        fun updatePostInformation(post: Post)

        fun setIsBusy(isBusy: Boolean)

        fun showImageSuccessSavedToGallery()

        fun updatePostImages(images: List<Image>)

        fun updateSimilarPosts(similarPosts: List<SimilarPost>)

        fun updatePostImage(image: File)

        fun updateImageDownloadProgress(progress: Int, maxProgress: Int)

        fun setEnableCreateComments()
    }
}